Integrate IPC pluggable infrastructure

araujol opened the pull request:
This commit integrates the IPC pluggable infrastructure
and an IPC (ZeroMQ) module that can be loaded using such
a mechanism.

It extends the vsm script to accept the '--ipc-module'
option to switch between the default stdin/stodut module
and the ZeroMQ IPC module.

It includes tests for the IPC module to check that the
plugin can be correctly loaded and that signals are
properly sent and processed by VSM.

Signed-off-by: Luis Araujo <luis.araujo@collabora.co.uk>

kov commented on tests.py:
```diff
@@ -81,5 +104,33 @@ def test_subclauses_arithmetic_booleans(self):
+        load_plugin('ipc.{}'.format(IPC_MODULE))
+
+    def test_simple0(self):
+        input_data = [('transmission_gear', '"reverse"')]
```
You can build the input data structure by processing the string. Like, you can split at =, strip spaces.

kov commented on tests.py:
```diff
@@ -81,5 +104,33 @@ def test_subclauses_arithmetic_booleans(self):
+
+    def test_simple0(self):
+        input_data = [('transmission_gear', '"reverse"')]
+        expected_output = ('car.backup', 'True')
```
Same here: you can build the strings back from the output, I think that would be better.

kov commented on tests.py:
```diff
@@ -81,5 +104,33 @@ def test_subclauses_arithmetic_booleans(self):
                 expected_output, False)
 
+
+class TestIPC(TestVSM):
```
This is just a different way of duplicating the tests, though, so it doesn't address the concern =).

kov commented on tests.py:
```diff
@@ -81,5 +104,33 @@ def test_subclauses_arithmetic_booleans(self):
+
 if __name__ == '__main__':
-    unittest.main()
+    IPC_MODULE=None
```
So what I was proposing was setting the flag as a variable on the class. So you'll do something like this:

```
TestVSM.use_ipc_module = None
unittest.main()
TestVSM.use_ipc_module = 'zeromq'
unittest.main()
```
The first call will run the tests with the default input/output, the second will use zeromq. Then in run_vsm you'll have:

```
if self.use_ipc_module:
   return self.run_with_ipc(…)
```

And in that function you can deal with all of the special cases such as breaking up the input string, putting the tuples back together, so that they can be properly compared.

For instance, this will take the input_data string the existing tests have and turn them into your list of tuples:

```
if self.use_ipc_module = 'zeromq':
    input_str = input_data
    input_data = []
    for line in input_str.strip().split('\n'):
        operands = map(str.strip, line.split('='))
        input_data.append(tuple(operands))
```

Let me know if you need help with that?

kov commented on ipc/zeromq.py:
```diff
@@ -0,0 +1,42 @@
+    socket.bind(IPC_SOCKET)
+
+    # Wait some time for subscriber to connect.
+    time.sleep(0.5)
```
We should have no sleeping for every send. This alone will kill performance in high throughput use cases.

kov commented on ipc/zeromq.py:
```diff
@@ -0,0 +1,42 @@
+
+def send(signal, value):
+    context = zmq.Context()
+    socket = context.socket(zmq.PUB)
```
You should only create the context and sockets (do send and receive need to be separate?) once - you can use one or more 'global' variables in the module for that.

kov commented on ipc/zeromq.py:
```diff
@@ -0,0 +1,42 @@
+    socket.send_pyobj((signal, value))
+
+
+def receive():
```
Same comments as above.

kov commented on tests.py:
```diff
@@ -8,12 +8,28 @@
+    "Function is overriden by plugin"
+
+def format_ipc_input(data):
+    return [ tuple(elm.split(' = ')) for elm in data.split('\n') ]
```
You should not use split(' = ') because signal=value (no spaces) is also allowed, if I remember correctly. That's why I had the strip.

kov commented on tests.py:
```diff
@@ -8,12 +8,28 @@
+    return [ tuple(elm.split(' = ')) for elm in data.split('\n') ]
+
+def format_ipc_output(data):
+    return (('', '') if data == '' else tuple(data.strip().split(' = ')))
```
I'd prefer if you translate the output that comes out of VSM into the string format for the output check.

araujol replied:
Any reason for this?

The way I see it, we shouldn't manipulate the output from the IPC message, but instead pass a correct expected IPC message to compare.

kov replied:
Because the expected output string is canonical (the 'official' standard), and it is easier to see differences in strings. In the future this will change a bit, since we will no longer use the signal emissions as 'log' and then you may have a check for the IPC emit and one for the string in the log. For now we're considering both the same, and the string is canonical.

kov commented on tests.py:
```diff
@@ -24,12 +40,29 @@ def run_vsm(self, name, input_data, expected_output, use_initial=True):
 
-        self.assertEqual(output.decode() , expected_output)
+            # Workaround for signals expecting ''
+            if expected_output == '':
```
I think you can send quit unconditionally? That won't be a problem for the other cases, right?

araujol replied:
If I send unconditionally, then the received message will always be the good bye ('', ''), message.

kov commented on vsm:
```diff
@@ -146,32 +150,57 @@ def process(state, signal, value):
+                exit(0)
+
+            signal, value = message
+            # 'quit' signal to close VSM end point.
```
Add a line before this comment, and also add a note that this is to simplify testing, please.

araujoluis committed:
Integrate IPC pluggable infrastructure

This commit integrates the IPC pluggable infrastructure
and an IPC (ZeroMQ) module that can be loaded using such
a mechanism.

It extends the vsm script to accept the '--ipc-module'
option to switch between the default stdin/stodut module
and the ZeroMQ IPC module.

It includes tests for the IPC module to check that the
plugin can be correctly loaded and that signals are
properly sent and processed by VSM.

Signed-off-by: Luis Araujo <luis.araujo@collabora.co.uk>

kov commented on tests.py:
```diff
@@ -25,22 +60,39 @@ def run_vsm(self, name, input_data, expected_output, use_initial=True):
+
+            for signal, value in format_ipc_input(input_data):
+                self._send(signal, value)
+            # Workaround for signals expecting ''
```
Empty line before the comment here.

kov commented on tests.py:
```diff
@@ -12,8 +12,43 @@
+             [ elm.split('=') for elm in data.split('\n') ] ]
+
+def format_ipc_output(data):
+    return (('', '') if data == '' else tuple(data.strip().split(' = ')))
```
I still don't like this, but it will probably change when the split between signals and logs is done, anyway, so I'm ok with keeping it.

kov reviewed (approved):
Just one nit. I think you should keep this in a branch now, close the task, and talk to Travis about how to integrate this after the split between logs and signals is completed in the tests.
